design/1.0.md

```diff
@@ -0,0 +1,84 @@
+I recently opened a Kubernetes Enhancement Proposal to introduce `go-flow-levee` during testing as a defense against accidental credential logging.
+See [KEP-1933](https://github.com/kubernetes/enhancements/pull/1936) details.
+Such integration demands a stable API and well-defined scope.
+
+I propose the following.
+Portions may already be implemented, but are included here as a "big picture."
+
+# 1.0 Targets
+
+## Basic taint propagation detection
+
+Given a spec defining sources and sinks (below), diagnostics are reported for
+any direct call to a sink that includes a source.  Additionally, we detect any
+taint propagation from a source to a sink within the scope of a single function.
+
+## Analysis Configuration
+
+Configuration will consist of the following key components:
+
+* Identification of sources, sinks, and sanitizers
+* Identification of analysis scope, e.g. to skip third-party or testing packages.
+* Specify diagnostics to be ignored, e.g false-positives
+
+### Identification of sources, sinks, and sanitizers
+
+Much of our work is directly with `ssa.Value`s and `ssa.Instruction`s.
+Whether those values or instructions represent one element or another,
+ the configuration to identify these members should be as uniform as possible.
+In that light, specification should follow:
+
+#### ValueSpecifier
+A ValueSpecifier marks an `ssa.Value`, e.g., for identification as a source or as the safe result of a sanitizer.
+Values should be specifiable via any or all of:
+
+* Type / Field: Package path and type name, optional field name
+* Field tags
+* Scope: local variable, free variable, or global
+* Context: within a specific package or function call.  See CallSpecifier below.
+* Is a reference
+* Is "reference-like," e.g., slices, maps, chans.
+
+#### CallSpecifier
+A CallSpecifier marks an `*ssa.Call`, e.g., as a sink or sanitizer.
+Calls should be specifiable via any or all of:
+
+* Symbol: package path, function name, optional receiver name.
+* Context: A CallSpecifier indicating  
+* Based on arguments (ValueSpecifier)
+* Based on return values (ValueSpecifier)
+
+## Inference
+
+Inference represents the scope of what a user does *not* need to configure.
+For instance, we can currently detect getter methods and mark as propagating
+a marked field without requiring any explicit configuration.
+
+If a source type is aliased, the alias is properly identified as equivalent as the aliased source type.
```

Comment:
For context, we already handle aliases e.g. `type MySource = core.Source`, but not named types e.g. `type MySource core.Source`. (Interestingly, in the short investigation I just did, the SSA was exactly the same for both cases).

Do we want named types to also count as "aliases"? Playing devil's advocate: what if a user defines a named type `type SafeSource core.Source` and uses it to explicitly differentiate between sources that have and have not been properly sanitized?

Reply:
> Do we want named types to also count as "aliases"?

Opinion: I think we should include named types.  The only difference between the two types would be the method set associated, which to me suggests that the kind of data stored is reasonably similar.  Also, mentioned below, I think it would be the easier way to handle the fact that field tags are naturally inherited in a way that a source specified by type path and name would not be.

Although...

> Playing devil's advocate: what if a user defines a named type type SafeSource core.Source and uses it to explicitly differentiate between sources that have and have not been properly sanitized?

I hadn't thought of that.  I think I might prefer an edge-case false-positive to a systematic false-negative, though.

Also and alternatively: this could be enabled/disabled in configuration.  And/or we could allow a user to include a safe-list for types that should not be considered sources, though that feels like it's getting a bit over-elaborate.

Reply:
I think you make a good case for named type inference.

I think the scenario I proposed is sufficiently unlikely that we should probably wait until we actually hit it before we think too deeply about how to handle it.

I think as a user I would expect a named type that aliases a Source to also be considered as Source.